### PR TITLE
[SC-288] Add pause support and external authorized contract

### PR DESCRIPTION
**Pull request by hexonaut**
Add the ability to pause/unpause + unfreeze as per requirements change. Also allow external contracts to do so (for multisig use case).

### CODE REVIEW

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Review by lucas-manuel — COMMENTED** [review]
Nice tests

File: src/interfaces/ISparkLendFreezerMom.sol [hunk 1]
```diff
@@ -56,6 +77,12 @@ interface ISparkLendFreezerMom {
      */
     function owner() external view returns (address);
 
+    /**
+     *  @dev    Returns if an address is authroized to trigger this mom (or not).
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/ISparkLendFreezerMom.sol:81]: fixed

File: src/interfaces/ISparkLendFreezerMom.sol [hunk 2]
```diff
@@ -72,25 +99,58 @@ interface ISparkLendFreezerMom {
...
+    function deny(address usr) external;
+
     /**********************************************************************************************/
     /*** Auth Functions                                                                         ***/
     /**********************************************************************************************/
 
     /**
      *  @dev   Function to freeze a specified market. Permissioned using the isAuthorized function
-     *         which allows the owner, the freezer contract itself, or the `hat` in the Chief
+     *         which allows the owner, a ward, the freezer contract itself, or the `hat` in the Chief
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/ISparkLendFreezerMom.sol:120]: done

File: src/interfaces/ISparkLendFreezerMom.sol [hunk 3]
```diff
@@ -72,25 +99,58 @@ interface ISparkLendFreezerMom {
...
+     *       which allows the owner, a ward, the freezer contract itself, or the `hat` in the Chief
+     *       to call the function. Note that the `authority` in this contract is assumed to be
+     *       the Chief in the MakerDAO protocol.
+     *  @param freeze A boolean indicating whether to freeze or unfreeze the market.
+     */
+    function freezeAllMarkets(bool freeze) external;
+
+    /**
+     *  @dev   Function to pause a specified market. Permissioned using the isAuthorized function
+     *         which allows the owner, a ward, the freezer contract itself, or the `hat` in the Chief
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/ISparkLendFreezerMom.sol:139]: done

File: test/IntegrationTests.t.sol [hunk 4]
```diff
@@ -6,14 +6,22 @@ import "forge-std/Test.sol";
 import { SafeERC20 } from "lib/erc20-helpers/src/SafeERC20.sol";
 import { IERC20 }    from "lib/erc20-helpers/src/interfaces/IERC20.sol";
 
+import { IExecuteOnceSpell }   from "src/interfaces/IExecuteOnceSpell.sol";
 import { SparkLendFreezerMom } from "src/SparkLendFreezerMom.sol";
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol:11]: done

File: src/spells/EmergencySpell_SparkLend_PauseSingleAsset.sol [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import { ISparkLendFreezerMom } from "src/interfaces/ISparkLendFreezerMom.sol";
+
+contract EmergencySpell_SparkLend_PauseSingleAsset is IExecuteOnceSpell {
+
+    address public immutable sparkLendFreezerMom;
+    address public immutable reserve;
+
+    bool public override executed;
+
+    constructor(address sparklendFreezerMom_, address reserve_) {
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/spells/EmergencySpell_SparkLend_PauseSingleAsset.sol:14]: done

File: test/IntegrationTests.t.sol [hunk 6]
```diff
@@ -47,39 +56,19 @@ contract IntegrationTestsBase is Test {
     SparkLendFreezerMom freezer;
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol:56]: done

File: test/IntegrationTests.t.sol [hunk 7]
```diff
@@ -410,19 +409,24 @@ contract FreezeSingleAssetSpellTest is IntegrationTestsBase {
...
     address[] public untestedReserves;
 
     function setUp() public override {
         super.setUp();
-        vm.prank(SPARK_PROXY);
-        aclManager.addRiskAdmin(address(freezer));
+
+        spell = new FreezeAllAssetsSpell(address(freezer));
+        isPauseSpell = false;
+        contractName = "FreezeAllAssetsSpell";
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol:423]: done

File: test/IntegrationTests.t.sol [hunk 8]
```diff
@@ -196,145 +227,113 @@ contract IntegrationTestsBase is Test {
...
 
     function setUp() public override {
         super.setUp();
-        vm.prank(SPARK_PROXY);
-        aclManager.addRiskAdmin(address(freezer));
+
+        // For the revert testing
+        spell = new FreezeSingleAssetSpell(address(freezer), WETH);
+        isPauseSpell = false;
+        contractName = "FreezeSingleAssetSpell";
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol:330]: done

File: test/IntegrationTests.t.sol [hunk 9]
```diff
@@ -508,3 +509,290 @@ contract FreezeAllAssetsSpellTest is IntegrationTestsBase {
...
+
+    address[] public untestedReserves;
+
+    function setUp() public override {
+        super.setUp();
+
+        // For the revert testing
+        spell = new PauseSingleAssetSpell(address(freezer), WETH);
+        isPauseSpell = true;
+        contractName = "PauseSingleAssetSpell";
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol:525]: done

File: test/IntegrationTests.t.sol [hunk 10]
```diff
@@ -508,3 +509,290 @@ contract FreezeAllAssetsSpellTest is IntegrationTestsBase {
...
+        }
+
+        // Undo all pauses and make sure that protocol is back to working
+        // as expected
+        for (uint256 i = 0; i < reserves.length; i++) {
+            address asset    = reserves[i];
+
+            vm.prank(SPARK_PROXY);
+            poolConfig.setReservePause(asset, false);
+        }
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol:694]: done

File: test/IntegrationTests.t.sol [hunk 11]
```diff
@@ -508,3 +509,290 @@ contract FreezeAllAssetsSpellTest is IntegrationTestsBase {
...
+
+        // For the revert testing
+        spell = new PauseSingleAssetSpell(address(freezer), WETH);
+        isPauseSpell = true;
+        contractName = "PauseSingleAssetSpell";
+    }
+
+    function test_pauseAssetSpell_allAssets() external {
+        vm.prank(SPARK_PROXY);
+        aclManager.addEmergencyAdmin(address(freezer));
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol]: moved it to parent contract and removed it for the one test

File: test/IntegrationTests.t.sol [hunk 12]
```diff
@@ -196,145 +227,113 @@ contract IntegrationTestsBase is Test {
...
+
+        // For the revert testing
+        spell = new FreezeSingleAssetSpell(address(freezer), WETH);
+        isPauseSpell = false;
+        contractName = "FreezeSingleAssetSpell";
     }
 
     function test_freezeAssetSpell_allAssets() external {
+        vm.prank(SPARK_PROXY);
+        aclManager.addRiskAdmin(address(freezer));
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol]: fixed

File: test/IntegrationTests.t.sol [hunk 13]
```diff
@@ -508,3 +509,290 @@ contract FreezeAllAssetsSpellTest is IntegrationTestsBase {
...
+        super.setUp();
+
+        spell = new PauseAllAssetsSpell(address(freezer));
+        isPauseSpell = true;
+        contractName = "PauseAllAssetsSpell";
+    }
+
+    function test_pauseAllAssetsSpell() external {
+        vm.prank(SPARK_PROXY);
+        aclManager.addEmergencyAdmin(address(freezer));
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply test/IntegrationTests.t.sol]: fixed

File: src/SparkLendFreezerMom.sol [hunk 14]
```diff
@@ -64,23 +61,48 @@ contract SparkLendFreezerMom is ISparkLendFreezerMom {
...
+            emit FreezeMarket(reserve, freeze);
+        }
+    }
+
+    function freezeMarket(address reserve, bool freeze) external override auth {
+        IPoolConfigurator(poolConfigurator).setReserveFreeze(reserve, freeze);
+        emit FreezeMarket(reserve, freeze);
+    }
+
+    function pauseAllMarkets(bool pause) external override auth {
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/SparkLendFreezerMom.sol:93]: We lose the event parity on `PauseMarket(...)` if we do that. The code is clearly the same, so I don't see much of a difference here. This function was added as convenience for multisig holders I believe. Also, don't know why they have a check for address 0.

[the remark this answers is not in the document]
Comment (lucas-manuel) [reply src/SparkLendFreezerMom.sol:93]: Fair makes sense

File: test/IntegrationTests.t.sol [hunk 15]
```diff
@@ -292,17 +305,17 @@ abstract contract ExecuteOnceSpellTests is IntegrationTestsBase {
 
     function test_cannotCallTwice() external {
         vm.prank(SPARK_PROXY);
-        if (isPauseSpell) aclManager.addEmergencyAdmin(address(freezer));
-        else aclManager.addRiskAdmin(address(freezer));
+        if (isPauseSpell) aclManager.addEmergencyAdmin(address(freezerMom));
+        else aclManager.addRiskAdmin(address(freezerMom));
```
Comment (lucas-manuel) [inline test/IntegrationTests.t.sol:309]: Can remove these lines now because of the setUp

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/IntegrationTests.t.sol (hunk 15 above)
Comment (hexonaut) [reply test/IntegrationTests.t.sol]: done

**Review by barrutko — APPROVED** [review]

**Pull request merged**